####

Index: trunk/daemon.c
===================================================================
--- trunk.orig/daemon.c
+++ trunk/daemon.c
@@ -814,6 +814,11 @@ static SVCXPRT *create_tcp_transport(uns
 	    fprintf(stderr, "Couldn't bind to tcp port %d\n", port);
 	    exit(1);
 	}
+	if (listen(sock, 10) < 0) {
+	    perror("listen");
+	    fprintf(stderr, "Couldn't listen to tcp port %d\n", port);
+	    exit(1);
+	}
     }
 
     transp = svctcp_create(sock, 0, 0);
