Index: git/libdnf/conf/OptionNumber.hpp
===================================================================
--- git.orig/libdnf/conf/OptionNumber.hpp
+++ git/libdnf/conf/OptionNumber.hpp
@@ -26,6 +26,7 @@
 #include "Option.hpp"
 
 #include <functional>
+#include <cstdint>
 
 namespace libdnf {
 
Index: git/libdnf/utils/sqlite3/Sqlite3.hpp
===================================================================
--- git.orig/libdnf/utils/sqlite3/Sqlite3.hpp
+++ git/libdnf/utils/sqlite3/Sqlite3.hpp
@@ -32,6 +32,7 @@
 #include <stdexcept>
 #include <string>
 #include <vector>
+#include <cstdint>
 
 class SQLite3 {
 public:
