Index: e2fsprogs-1.41.14/lib/et/compile_et.sh.in
===================================================================
--- e2fsprogs-1.41.14.orig/lib/et/compile_et.sh.in
+++ e2fsprogs-1.41.14/lib/et/compile_et.sh.in
@@ -51,12 +51,12 @@ if test -f ${BASE}.h && cmp -s ${BASE}.h
     rm -f ${BASE}.h.$$
 else
     mv -f ${BASE}.h.$$ ${BASE}.h
-    chmod -w ${BASE}.h
+    chmod a-w ${BASE}.h
 fi
 $AWK -f "${DIR}/et_c.awk" "outfile=${BASE}.c.$$" "outfn=${BASE}.c" "$ROOT.et"
 if test -f ${BASE}.c && cmp -s ${BASE}.c.$$ ${BASE}.c ; then
     rm -f ${BASE}.c.$$
 else
     mv -f ${BASE}.c.$$ ${BASE}.c
-    chmod -w ${BASE}.c
+    chmod a-w ${BASE}.c
 fi
Index: e2fsprogs-1.41.14/lib/ss/mk_cmds.sh.in
===================================================================
--- e2fsprogs-1.41.14.orig/lib/ss/mk_cmds.sh.in
+++ e2fsprogs-1.41.14/lib/ss/mk_cmds.sh.in
@@ -57,6 +57,6 @@ if grep "^#__ERROR_IN_FILE" "${TMP}" > /
 else
	rm -f "${BASE}.c"
	mv -f "${TMP}" "${BASE}.c"
-	chmod -w "${BASE}.c"
+	chmod a-w "${BASE}.c"
	exit 0
 fi
