Upstream-Status: Pending

Index: strace-6.15/src/v4l2.c
===================================================================
--- strace-6.15.orig/src/v4l2.c
+++ strace-6.15/src/v4l2.c
@@ -14,6 +14,7 @@
 #include DEF_MPERS_TYPE(kernel_v4l2_buffer_time32_t)
 #include DEF_MPERS_TYPE(kernel_v4l2_event_t)
 #include DEF_MPERS_TYPE(kernel_v4l2_timeval_t)
+#include DEF_MPERS_TYPE(kernel_v4l2_exportbuffer_t)
 #include DEF_MPERS_TYPE(struct_v4l2_clip)
 #include DEF_MPERS_TYPE(struct_v4l2_create_buffers)
 #include DEF_MPERS_TYPE(struct_v4l2_ext_control)
@@ -583,7 +584,8 @@ print_v4l2_buffer_contents(struct tcb *c
 	PRINT_FIELD_U(*b, bytesused);
 	tprint_struct_next();
 	PRINT_FIELD_V4L2_BUFFER_FLAGS(*b, flags);
-	if (code == VIDIOC_DQBUF) {
+	if ((code == VIDIOC_DQBUF && b->type == V4L2_BUF_TYPE_VIDEO_CAPTURE) ||
+	    (code == VIDIOC_QBUF  && b->type == V4L2_BUF_TYPE_VIDEO_OUTPUT)) {
 		tprint_struct_next();
 		PRINT_FIELD_V4L2_TIMEVAL(*b, timestamp);
 	}
@@ -681,6 +683,52 @@ print_v4l2_buffer_time32(struct tcb *con
 #endif
 
 static int
+print_v4l2_expbuf(struct tcb *const tcp, const unsigned int code,
+		  const kernel_ulong_t arg)
+{
+	struct v4l2_exportbuffer b;
+
+	if (umove_or_printaddr(tcp, arg, &b)) {
+		tprint_arg_next();
+		return RVAL_IOCTL_DECODED;
+	}
+
+	if (entering(tcp)) {
+		tprint_arg_next();
+
+		tprint_struct_begin();
+		printxval(v4l2_buf_types, b.type, "V4L2_BUF_TYPE_???");
+		tprint_struct_next();
+		PRINT_FIELD_U(b, index);
+		tprint_struct_next();
+		PRINT_FIELD_U(b, plane);
+		tprint_struct_next();
+		tprint_open_modes(b.flags);
+
+		if (b.fd != 0) {
+			tprint_struct_next();
+			printfd(tcp, b.fd);
+		}
+
+		if (!IS_ARRAY_ZERO(b.reserved)) {
+			tprint_struct_next();
+			PRINT_FIELD_ARRAY(b, reserved, tcp,
+					  print_xint_array_member);
+		}
+
+		tprint_struct_end();
+		return 0;
+	} else if (!syserror(tcp)) {
+		tprint_value_changed();
+		tprint_struct_begin();
+		printfd(tcp, b.fd);
+		tprint_struct_end();
+	}
+
+	return RVAL_IOCTL_DECODED;
+}
+
+static int
 print_v4l2_framebuffer(struct tcb *const tcp, const kernel_ulong_t arg)
 {
 	struct_v4l2_framebuffer b;
Index: strace-6.15/src/kernel_v4l2_types.h
===================================================================
--- strace-6.15.orig/src/kernel_v4l2_types.h
+++ strace-6.15/src/kernel_v4l2_types.h
@@ -101,6 +127,15 @@ typedef struct {
 	uint32_t				reserved[8];
 } kernel_v4l2_event_t;
 
+typedef struct {
+	uint32_t		type; /* enum v4l2_buf_type */
+	uint32_t		index;
+	uint32_t		plane;
+	uint32_t		flags;
+	uint32_t		fd;
+	uint32_t		reserved[11];
+} kernel_v4l2_exportbuffer_t;
+
 /* Removed by Linux kernel commit v3.6-rc1~28^2~240.  */
 # define V4L2_BUF_FLAG_INPUT	0x0200
 
